examples/ls-files.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * libgit2 "ls-files" example - shows how to view all files currently in the index
+ *
+ * Written by the libgit2 contributors
+ *
+ * To the extent possible under law, the author(s) have dedicated all copyright
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * You should have received a copy of the CC0 Public Domain Dedication along
+ * with this software. If not, see
+ * <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+
+#include "common.h"
+
+/**
+ * This example demonstrates the libgit2 index APIs to roughly
+ * simulate the output of `git ls-files`.
+ * `git ls-files` has many options and this currently does not show them.
+ * 
+ * `git ls-files` base command shows all paths in the index at that time.
+ * This includes staged and committed files, but unstaged files will not display.
+ * 
+ * This currently supports:
+ * 	- The --error-unmatch paramter with the same output as the git cli
+ *  - default ls-files behavior
+ * 
+ * This currently does not support:
+ * 	- anything else
+ * 
+ */
+
+#define MAX_FILES 64
+
+typedef struct ls_options {
+	int error_unmatch;
+	char *files[MAX_FILES];
+	int file_count;
+} ls_options;
+
+static void usage(const char *message, const char *arg);
+void parse_options(ls_options *opts, int argc, char *argv[]);
+int print_error_unmatch(ls_options *opts, git_index *index);
+
+int main(int argc, char *argv[]) {
+	ls_options opts;
+	git_repository *repo;
+	git_index *index;
+	const git_index_entry *entry;
+	size_t entry_count;
+	size_t i = 0;
+	int error;
+
+	parse_options(&opts, argc, argv);
+
+	/* we need to initialize libgit2 */
+	git_libgit2_init();
+
+	/* we need to open the repo */
+	if ((error = git_repository_open_ext(&repo, ".", 0, NULL)) != 0)
+		goto cleanup;
+
+	/* we need to load the repo's index */
+	if ((error = git_repository_index(&index, repo)) != 0)
+		goto cleanup;
+
+	/* if the error_unmatch flag is set, we need to print it differently */
+	if (opts.error_unmatch) {
+		error = print_error_unmatch(&opts, index);
+		goto cleanup;
+	}
```

Comment:
I think you've got a slight misconception of `error_unmatch`. What you do here is print out different things based on whether `error_unmatch` is set or not. But in fact, in both cases we'd like to either list the complete index if we have no explicit file list, or in the other case just the files. So _what_ is printed should only be decided based on whether the user explicitly passed files into the program or not. `error_unmatch` is then only used in the case where we have paths to check if the path actually exists.